src/common/com/intellij/plugins/haxe/model/type/SpecificHaxeClassReference.java

```diff
@@ -182,4 +183,167 @@ public ResultHolder access(String name, HaxeExpressionEvaluatorContext context)
     }
     return null;
   }
+
+  Set<SpecificHaxeClassReference> getCompatibleTypes() {
+    Set<SpecificHaxeClassReference> result = context.getUserData(COMPATIBLE_TYPES_KEY);
+    if (result == null) {
+      processedElements.get().clear();
+      result = getCompatibleTypesInternal();
+      result.add(this);
+      context.putUserData(COMPATIBLE_TYPES_KEY, result);
+    }
+    return result;
+  }
+
+  Set<SpecificHaxeClassReference> getInferTypes() {
+    Set<SpecificHaxeClassReference> result = context.getUserData(INFER_TYPES_KEY);
+    if (result == null) {
+      processedElements.get().clear();
+      result = getInferTypesInternal();
+      context.putUserData(INFER_TYPES_KEY, result);
+    }
+    return result;
+  }
+
+  private Set<SpecificHaxeClassReference> getCompatibleTypesInternal() {
+    final Stack<HaxeClass> stack = processedElements.get();
+    final HaxeClassModel model = getHaxeClassModel();
+    final HaxeGenericResolver genericResolver = getGenericResolver();
+
+    final Set<SpecificHaxeClassReference> list = new HashSet<>();
+
+    if (model == null) return list;
+    if (stack.contains(model.haxeClass)) return list;
+    stack.push(model.haxeClass);
+
+    if (!model.isAbstract()) {
+      if (model.haxeClass instanceof HaxeTypedefDeclaration) {
+        SpecificHaxeClassReference type = ((AbstractHaxeTypeDefImpl)model.haxeClass).getTargetClass(genericResolver);
+        if (type != null) {
+          list.add(type);
+          list.addAll(type.getCompatibleTypes());
+        }
```

Comment:
Well, they are not exactly the same. I'm not sure if it could be encapsulated in way of better looking. Right now I can't say it's possible to make it better.